### PR TITLE
Allow connecting/disconnecting from networks in UI, closes #631

### LINKED ISSUE #631: When disconnected, network menu still shows "disconnect" option
When a network is disconnected, lounge tells you to use `/connect` to reconnect, so the menu should provide a connect option too instead of disconnect.

### PR DESCRIPTION
**Pull request by MaxLeiter**
Few thoughts that need to be worked out:

1. Not sure how I feel about icons
<table>
  <tr>
    <td>
screen shot 2018-03-17 at 2 06 24 pm
</td>
    <td>
screen shot 2018-03-17 at 2 06 29 pm

</td>
  </tr>
</table>

2. Should status be saved server-side? Add a `connected` field to `Network`?
3. More I'm sure I forgot
4. Need to deal with add-channel UI when disconnected. Thoughts?

### CODE REVIEW

**Comment by christer88** [comment]
I guess this PR will close https://github.com/thelounge/thelounge/issues/1112 because we now have "Disconnect" to only disconnect, and "Remove" to remove network from json config file? If so, thanks @MaxLeiter!

**Comment by McInkay** [comment]
My poor #1878

**Comment by creesch** [comment]
> Need to deal with add-channel UI when disconnected. Thoughts?

Leave channels with logging in place, disable input field for channels on disconnected networks.

**Comment by xPaw** [comment]
There's no need to disable the input because lounge will join it after connecting.

**Comment by MaxLeiter** [comment]
Ready to go?

**Comment by McInkay** [comment]
Oh please... Mine first.

**Comment by MaxLeiter** [comment]
Fine with me, will be a simple rebase

**Comment by McInkay** [comment]
Well let's see what @astorije says about mine after my refactor earlier today. It's just rebasing mine on top of changes does ideas the likelihood of regressions. I'm not going to ask you to wait weeks.

**Comment by Haarolean** [comment]
Would be nice to be able to change network settings via UI as well. 
https://github.com/thelounge/thelounge/issues/179 is related.

**Comment by McInkay** [comment]
@Haarolean This is unrelated to the PR, please keep relevant discussion to their own tickets, as you have commented on the other issue. Thanks.

**Comment by astorije** [comment]
I don't think "Disconnect" is intentional here, right?

screen shot 2018-03-04 at 20 35 28

**Comment by astorije** [comment]
Do we need both entries "Disconnect" and "Remove" (which also disconnects)?

**Comment by MaxLeiter** [comment]
@astorije: Remove deletes, disconnect disconnects

**Comment by astorije** [comment]
Yes I understand that. Remove also disconnects (right?). But, do we actually need that?

I sort of understand the need for `Connect` (actually, `Reconnect`?) because connection with server can be lost from time to time, and there are cases when we don't automatically reconnect (though we should limit this as much as we can), but why "Disconnect" on top of "Remove"?
It seems to me that the main purpose of TL is to stay connected at all time, reconnect whenever possible, etc. so I'm just trying to understand the need behind this, in case we're trying to solve the wrong problem.

**Comment by MaxLeiter** [comment]
There are times where you want to disconnect so you can reconnect, and there are times when you want to delete the network. Sometimes you want to save the network info but temporally leave it. Pretty common in clients, afaik

**Comment by astorije** [comment]
What about a "Reconnect" instead that disconnects + reconnects? That seems to me to be the end-goal here.

**Comment by McInkay** [comment]
> What about a "Reconnect" instead that disconnects + reconnects? That seems to me to be the end-goal here.

It's the end goal for some of it, sure, but there is definitely a use case for an irc client that you maybe don't always want to be connected to, and making "Disconnect" and "Connect" instead of "Reconnect" allows both use cases.

**Comment by christer88** [comment]
I agree with @McInkay!

**Comment by MaxLeiter** [comment]
Updated links to FA 5, rebased to master, updated screenshots

**Comment by McInkay** [comment]
This will need to be refactored now that #1878 is merged, and use the new structure for adding context menu items. If you have any questions about how this works, or anything is unclear about what I've changed then give me a shout @MaxLeiter

**Comment by MaxLeiter** [comment]
@McInkay: working on updating to master.
2 questions
1) Where should I set `connected`? Currently, I set it in `check()`.
2) `displayName` does not seem to be updating. (As in, always shows "Connected"). Fixing number 1 properly may fix this inherently (like if I make two different `ContextMenuItem`'s)


``` javascript
	let connected;
	contextMenuFactory.addContextMenuItem({
		check: (target) => {
			// not ideal to have the line below here. Not sure where else to place it.
			connected = !target.closest("section").hasClass("not-connected");
			console.log(target.closest("section").hasClass("not-connected"));
			return target.hasClass("lobby");
		},
		className: connected ? "disconnect" : "connect",
		displayName: connected ? "Disconnect" : "Connect",
		data: (target) => target.data("id"),
		callback: (itemData) => {
			if (!connected) {
				socket.emit("input", {
					target: itemData,
					text: "/connect",
				});
			} else {
				socket.emit("input", {
					target: itemData,
					text: "/disconnect",
				});
			}
		},
	});
```

**Comment by McInkay** [comment]
@MaxLeiter 

> Where should I set connected? Currently, I set it in check()

I think you'll need to set it onto the object, otherwise the scope won't work? Maybe I'm wrong.

> displayName does not seem to be updating.

Yeah, this is because you are passing through the string, not a function. Change it to this:

```
className: () => connected ? "disconnect" : "connect",
displayName: () => connected ? "Disconnect" : "Connect",
```

This way the class and display name are recalculated everytime it creates a menu item, rather than just when you create it.

Thanks for the advice, I'll make sure to put this more clearly into the docs when I write them.

Hope that makes sense.

**Comment by MaxLeiter** [comment]
Rebased + updated, running into a slight issue with `className` though -- should not need arrow notation for it to function

**Comment by McInkay** [comment]
> running into a slight issue with className though -- should not need arrow notation for it to function

Yes, you absolutely should.

If you are passing through `connected ? "Disconnect" : "Connect"` to the function, that is processed by Javascript and sent as a String. Therefore it'll always be the same. You need to send a function for it to ever change.

File: client/js/lounge.js [hunk 1]
```diff
@@ -330,14 +330,48 @@ $(function() {
...
+		return "Connect";
+	};
+
+	let clickedNetwork;
+	contextMenuFactory.addContextMenuItem({
+		check(target) {
+			clickedNetwork = target;
+			return target.hasClass("lobby");
+		},
+		className: () => isConnected(clickedNetwork) ? "connect" : "disconnect", // eslint-disable-line no-confusing-arrow
```
Comment (astorije) [inline client/js/lounge.js]: I'm not fan of disabling stuff like this. Use the following instead:

```js
className() {
  return isConnected(clickedNetwork) ? "connect" : "disconnect";
},
```

**Comment by MaxLeiter** [comment]
Yeah just haven’t got to it yet

**Comment by MaxLeiter** [comment]
Should be good for reviews

**Comment by astorije** [comment]
Thanks @MaxLeiter!

File: client/js/lounge.js [hunk 2]
```diff
@@ -330,14 +330,50 @@ $(function() {
...
+	const getToggleConnectionDisplay = (target) => {
+		if (isConnected(target)) {
+			return "Disconnect";
+		}
+
+		return "Connect";
+	};
+
+	let clickedNetwork;
+	contextMenuFactory.addContextMenuItem({
```
Comment (astorije) [inline client/js/lounge.js]: @xPaw raised a good point, could you split this into 2 items? I believe you'd have your `isConnected(target)` in `check`. That would simplify things like `className` 😅

Comment (xPaw) [reply client/js/lounge.js]: Also move all this code into context menu factory.

Comment (MaxLeiter) [reply client/js/lounge.js]: That complicates things like `isConnected()`, will need to play with this

**Comment by MaxLeiter** [comment]
Okay, split up connect/disconnect and moved to factory.

Also moved other context code left-over from lounge.js -> factory, but that required moving `closeChan` to `utils.js` so the Factory could use it. I don't care if we leave the close chan context item in `lounge.js` or not, just let me know.

**Pull request merged**